chain/client/src/stateless_validation/state_witness_tracker.rs

```diff
@@ -0,0 +1,201 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use lru::LruCache;
+use std::hash::Hash;
+
+use crate::metrics;
+use near_async::time::Clock;
+use near_primitives::sharding::ChunkHash;
+use near_primitives::stateless_validation::{ChunkStateWitness, ChunkStateWitnessAck};
+
+/// This is currently used for metrics computation, so we do not need to keep all the
+/// witnesses, so we set some fixed capacity for the witnesses tracked and discard others.
+const CHUNK_STATE_WITNESS_MAX_RECORD_COUNT: usize = 50;
+
+/// Refers to a state witness sent to a chunk producer. Used to locate incoming acknowledgement
+/// messages back to the timing information of the originating witness.
+#[derive(Debug, Clone, Hash, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+pub struct ChunkStateWitnessKey {
+    /// Hash of the chunk for which the state witness was generated.
+    pub chunk_hash: ChunkHash,
+}
+
+impl ChunkStateWitnessKey {
+    pub fn new(witness: &ChunkStateWitness) -> Self {
+        Self { chunk_hash: witness.inner.chunk_header.chunk_hash() }
+    }
+}
+
+pub struct ChunkStateWitnessRecord {
+    /// Size of the witness in bytes.
+    pub witness_size: usize,
+    /// Number of validators that the witness is sent to.
+    pub num_validators: usize,
+    /// Timestamp of when the chunk producer sends the state witness.
+    pub sent_timestamp: near_async::time::Instant,
+}
+
+/// Tracks a collection of state witnesses sent from chunk producers to validators.
+/// This is currently used to calculate the round-trip time of sending the witness and
+/// getting the ack message back, where the ack message is used as a proxy for the endorsement
+/// message from validators to block producers to make an estimate of network time for sending
+/// witness and receiving the endorsement.
+pub struct ChunkStateWitnessTracker {
+    witnesses: LruCache<ChunkStateWitnessKey, ChunkStateWitnessRecord>,
+    clock: Clock,
+}
+
+impl ChunkStateWitnessTracker {
+    pub fn new(clock: Clock) -> Self {
+        Self { witnesses: LruCache::new(CHUNK_STATE_WITNESS_MAX_RECORD_COUNT), clock }
+    }
+
+    /// Adds a new witness message to track.
+    pub fn record_witness_sent(
+        &mut self,
+        witness: &ChunkStateWitness,
+        witness_size_in_bytes: usize,
+        num_validators: usize,
+    ) -> () {
+        let key = ChunkStateWitnessKey::new(witness);
+        tracing::debug!(target: "state_witness_tracker", "Sent state witness: {:?} of {} bytes",
```

Comment:
```suggestion
        tracing::trace!(target: "state_witness_tracker", "Sent state witness: {:?} of {} bytes",
```

fairly arbitrary suggestion, but this definitely looks more like an execution trace event more so than a debug event.